### ISSUE TITLE
Incorrect data

### ISSUE THREAD

**Issue by davidanthoff** [tk3369/SASLib.jl#8]
I tried to read https://github.com/davidanthoff/StatFiles.jl/blob/master/test/types.sas7bdat, and I essentially get incorrect data for every single field :) You can see the expected values [here](https://github.com/davidanthoff/StatFiles.jl/blob/master/test/runtests.jl#L14).

**Comment by tk3369** [comment]
Good catch @davidanthoff!  It puzzles me as my other test files are loaded correctly.  I'm not sure what's wrong yet but at least I've proven that Pandas returns wrong values as well so it's not a porting issue.

```
$ python
Python 3.6.3 |Anaconda custom (64-bit)| (default, Oct  6 2017, 12:04:38) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from pandas.io.sas import read_sas
>>> read_sas("types.sas7bdat")
          vfloat        vdouble          vlong           vint          vbyte  \
0   4.276981e+86  1.780060e-307  1.780059e-307   1.187048e-66  5.432309e-312   
1  5.314056e-315  1.792228e-307  1.792227e-307   1.444739e-42  5.554059e-313   
2  2.121990e-314  -9.946247e+86  -9.945706e+86  9.774603e-150  1.916206e-312   

   vstring  
0  b'\x00'  
1  b'\x00'  
2  b'\x00'  
```

Results from SASLib as a reference:

```
julia> df
3×6 DataFrames.DataFrame
│ Row │ vbyte │ vdouble      │ vfloat       │ vint        │ vlong        │ vstring │
├─────┼───────┼──────────────┼──────────────┼─────────────┼──────────────┼─────────┤
│ 1   │ 0.0   │ 1.78006e-307 │ 4.27698e86   │ 1.18695e-66 │ 1.78006e-307 │ \0      │
│ 2   │ 0.0   │ 1.79223e-307 │ 5.31406e-315 │ 1.44474e-42 │ 1.79223e-307 │ \0      │
│ 3   │ 0.0   │ -9.94625e86  │ 2.12199e-314 │ 9.7746e-150 │ -9.94525e86  │ \0      │
```

**Comment by davidanthoff** [comment]
Hehe :) I guess this is testing some of the more obscure data types...

**Comment by tk3369** [comment]
Found the problem.  It's probably the same issue with all other numbers.  Let's say the number is represented in 8 bytes.  We know that BigEndian and LittleEndian systems order them differently.  In the case of types.sas7bdat, it's LittleEndian.  However, it's working off a 32-bit boundary (maybe it was created from a 32-bit system?)  Hence the byte swap routine cannot just reverse the 8 bytes but rather having to do them 4 bytes at a time.   The following code proves it.

```
julia> # from types.sas7bdat 
       # vfloat column, 1st row => this value should be 3.14 even
       v = UInt8[0xb8, 0x1e, 0x09, 0x40, 0x1f, 0x85, 0xeb, 0x51]
8-element Array{UInt8,1}:
 0xb8
 0x1e
 0x09
 0x40
 0x1f
 0x85
 0xeb
 0x51

julia> # from SASLib.jl
       function convertint64(a::UInt8,b::UInt8,c::UInt8,d::UInt8,e::UInt8,f::UInt8,g::UInt8,h::UInt8)
           (UInt64(a) << 56) | (UInt64(b) << 48) | 
           (UInt64(c) << 40) | (UInt64(d) << 32) | 
           (UInt64(e) << 24) | (UInt64(f) << 16) | 
           (UInt64(g) << 8)  |  UInt64(h)
       end
convertint64 (generic function with 1 method)

julia> w = convertint64(v...)
0xb81e09401f85eb51

julia> # for little endian 
       w2 = bswap(w)
0x51eb851f40091eb8

julia> # what SASLib.jl currently returns, which is wrong
       reinterpret(Float64, w2)
4.276981090307826e86

julia> # it turns out that byte alignment isn't right
       # try this instead => swap 32-bit values 
       w3 = ((0x00000000ffffffff & w2) << 32) | (w2 >> 32)
0x40091eb851eb851f

julia> # hooray!
       reinterpret(Float64, w3)
3.14
```

**Comment by davidanthoff** [comment]
Ah, yeah, that sounds plausible. I guess the question is whether ``types.sas7bdat`` actually conforms to the file spec? Or am I using a file that is just weird and shouldn't read successfully? I don't even remember where I got it from, it probably originated with ReadStat or something like that...

**Comment by tk3369** [comment]
Given that SAS knows how to read the file correctly, I would consider this a bug.  My next hurdle is how to identify when to swap the bytes at the 32-bit word boundary.  There's got to be a magic number somewhere.

**Comment by tk3369** [comment]
Fixed by commit 36b8ef5 in develop branch

**Comment by tk3369** [comment]
For the record, the comment below is incorrect.  The real problem is that the reader was misaligned by exactly 4 bytes causing it to only read half of the data in the exact opposite location. Apparently files created by STAT_TRANSFER require some special handling.

> However, it's working off a 32-bit boundary (maybe it was created from a 32-bit system?) Hence the byte swap routine cannot just reverse the 8 bytes but rather having to do them 4 bytes at a time. The following code proves it.
